src/goose/brokers/redis/batch.clj

```diff
@@ -0,0 +1,164 @@
+        (car/lpush (:ready-queue job) job)))))
+
+;;; When fetching a hash from Redis, primitive data types
+;;; like integer, boolean and keyword get converted to string.
+;;; We restore data types to avoid any type-casting errors.
+(defn- restore-data-types
+  [{:keys [linger-sec total status created-at] :as batch}]
+  (assoc batch
+    :linger-sec (Integer/valueOf linger-sec)
+    :total (Integer/valueOf total)
+    :status (keyword status)
+    :created-at (Long/valueOf created-at)))
+
+(defn get-batch
+  [redis-conn id]
+  (let [{:keys [batch-hash enqueued-set retrying-job-set success-set dead-set]} (batch-keys id)
+        [batch enqueued retrying success dead] (redis-cmds/wcar*
+                                                 redis-conn
+                                                 (car/parse-map (car/hgetall batch-hash) :keywordize)
+                                                 (car/scard enqueued-set)
+                                                 (car/scard retrying-job-set)
+                                                 (car/scard success-set)
+                                                 (car/scard dead-set))]
+    (when (not-empty batch)
+      (assoc (restore-data-types batch)
+        :enqueued enqueued
+        :retrying retrying
+        :success success
+        :dead dead))))
+
+(defn- record-metrics
+  [{:keys [metrics-plugin]}
+   {:keys [execute-fn-sym]}
+   {:keys [id queue created-at]}
+   completion-status]
+  (when (m/enabled? metrics-plugin)
+    (let [completion-time (- (u/epoch-time-ms) created-at)
+          tags {:batch-id id :execute-fn-sym execute-fn-sym :queue queue}]
+      (m/increment metrics-plugin (m/format-batch-status completion-status) 1 tags)
+      (m/timing metrics-plugin m/batch-completion-time completion-time tags))))
+
+(defn- enqueue-callback
+  [redis-conn
+   {:keys [id ready-queue linger-sec] :as batch}
+   completion-status]
+  (let [{:keys [batch-hash enqueued-set retrying-set success-set dead-set]} (batch-keys id)
+        callback (batch/new-callback batch id completion-status)]
+    (redis-cmds/atomic
+      redis-conn
+      (car/multi)
+      ;; Enqueue callback to front of queue.
+      (car/rpush ready-queue callback)
+      ;; Update batch status to reflect completion status.
+      (car/hset batch-hash :status completion-status)
+      ;; Terminal job execution marks completion of a batch, NOT callback execution.
+      ;; Clean-up batch after enqueuing callback.
+      (car/expire batch-hash linger-sec "NX")
+      (car/expire enqueued-set linger-sec "NX")
+      (car/expire retrying-set linger-sec "NX")
+      (car/expire success-set linger-sec "NX")
+      (car/expire dead-set linger-sec "NX"))))
+
+(defn- mark-batch-completion
+  [{:keys [redis-conn] :as opts} job batch-id completion-status]
+  ;; If a job is executed after a batch has been deleted,
+  ;; `when-let` guards against nil return from `get-batch`.
+  (when-let [batch (get-batch redis-conn batch-id)]
```

Comment:
Can you add a log when the batch is not found but `mark-batch-completion` is called. This should happen very rarely if at all, but an error log will help in case it does.

Reply:
Added a warning log, since it is not an error for a job to execute when batch is missing from broker. Can occur in 2 cases: `orphaned-job re-running after batch metadata clean-up` or `an in-progress job executing post batch/delete API call.`